### ISSUE TITLE
account.declare do not like compressed Cairo programs

### ISSUE THREAD

**Issue by PhilippeR26** [0xs34n/starknet.js#588]
**Describe the bug**
I have many problems to use Account.declare in v5.5.0.
1. `Account.declare` do not accept `CairoContract` type, only `string|CompiledContract`. It means that `SierraContractClass` is not authorized. 
So, use of Provider.getClassAt() or Provider.getClassByHash() with a Cairo1 class do not allow to use the result in `account.declare`. 
2. If I try to declare a compressed Cairo 0(LegacyContractClass), my script is now  failing. json-bigint library has been replaced by lossless-json : it generate a crash in parse  ; lossless-json seems to not accept LegacyContractClass.program. Also very problematic.

**To Reproduce**
```typescript
async function main() {

    //initialize Provider with testnet2
    const providerTestnet2 = new Provider({ sequencer: { network: constants.NetworkName.SN_GOERLI2 } });
    console.log("Connected to testnet2.");

    //initialize Provider with devnet
    const providerDevnet = new Provider({ sequencer: { baseUrl: "http://127.0.0.1:5050" } });
    console.log('Connected to devnet.');

    // initialize existing predeployed account 0 of Devnet
    const privateKey = "0xe3e70682c2094cac629f6fbed82c07cd";
    const accountAddress = "0x7e00d496e324876bbc8531f2d9a82bf154d1a04a50218ee74cdd372f75a551a";
    const account0 = new Account(providerDevnet, accountAddress, privateKey);
    console.log('✅ OZ predeployed account 0 connected in devnet.');


    // recover & save contract
    const compressedContractProxy = await providerTestnet2.getClassByHash("0x025ec026985a3bf9d0cc1fe17326b245dfdc3ff89b8fde106542a3ea56c5a918");
    fs.writeFileSync('./compiledContracts/ArgentXproxy_0_2_3_1_compressed.json', JSON.stringify(compressedContractProxy));

    // declare in devnet
    if ('sierra_program' in compressedContractProxy) { throw new Error("Cairo 1 compressed not authorized in Starknet.js v5.5.0") }
    const progCompressed = compressedContractProxy.program;

    // 🚨🚨🚨 Fail in v5.5.0 ; crash in lossless-json library
    const declareResponse = await account0.declare({ contract: progCompressed });
    console.log("clashHash =", declareResponse.class_hash);

}
main()
    .then(() => process.exit(0))
    .catch((error) => {
        console.error(error);
        process.exit(1);
    });
```

**Expected behavior**
Account.declare to accept `CairoContract`, and process it correctly.

**Screenshots**
```
SyntaxError: JSON value expected but got 'H' at position 0
    at expectValue (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/lossless-json/lib/esm/parse.js:201:13)
    at Object.t.parse (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/lossless-json/lib/esm/parse.js:36:3)
    at parse (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/starknet/src/utils/json.ts:12:47)
    at isSierra (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/starknet/src/utils/contract.ts:7:59)
    at extractContractHashes (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/starknet/src/utils/contract.ts:16:7)
    at Account.declare (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/node_modules/starknet/src/account/default.ts:338:36)
    at main (/home/edmond/Documents/starknet/starknet.js-workshop-typescript/src/scripts/mainnet/2b.recoverCompiledContract.ts:50:44)

```
**Desktop (please complete the following information):**

- Browser & version [e.g. chrome, safari, webworker]
- Node version [e.g. 16.0.1]
- Starknet.js version : V5.5.0
- Network [devnet, testnet] devnet 0.5.0

**Comment by penovicp** [comment]
The mentioned parsing issue would arise regardless of which underlying dependency/module is used. The `getClassByHash` method of the `SequencerProvider` class adjusts the response it receives so it can provide the same output as the equivalent method of the `RpcProvider` class, it does so by compressing the `program` property ([link](https://github.com/0xs34n/starknet.js/blob/v5.5.0/src/utils/provider.ts#L41)). 

So to make the `program` property value usable it would first need to be decompressed with something like:
```
import * as pako from 'pako'

const decompress = (base64: string) => Buffer.from(pako.ungzip(Buffer.from(base64, 'base64'))).toString()
```

With that being said, should extracting `program` as a subset and reusing it like that through `declare` even be feasible? Wouldn't it have an improper/insufficient structure?

**Comment by PhilippeR26** [comment]
Tests made with a Cairo 0 contract (Cairo 1 has a different program format) :
It works when I read a contract from devnet, reset devnet and redeclare the contract with the json collected.
From mainnet/testnet to devnet (the most interesting case), I have a signature error when declaring. I don"t know how to solve the problem.
My code : 
```typescript
const decompress = (base64: string) => Buffer.from(pako.ungzip(Buffer.from(base64, 'base64'))).toString();
    const classHash = "0x74953440d81a1b2deb04905bec239d8879f98bad63350ec550a07a7ff13dbc0";
    const compressedContract = await providerDevnet.getClassByHash(classHash);
    if (("program" in compressedContract) && compressedContract.abi) { // Cairo 0
        const decompressedProgram = json.parse(decompress(compressedContract.program));
        fs.writeFileSync('./src/scripts/devnet/erc20_program.json', json.stringify(decompressedProgram, undefined, 2));
        const compiledContract: LegacyCompiledContract = {
            entry_points_by_type: compressedContract.entry_points_by_type,
            abi: compressedContract.abi,
            program: decompressedProgram
        }
        fs.writeFileSync('./src/scripts/devnet/erc20_contract.json', json.stringify(compiledContract, undefined, 2));
    } else {
        // Cairo 1
    }
```
Then to redeclare after devnet reset :
```typescript
const compiledTest = json.parse(fs.readFileSync("./src/scripts/devnet/erc20_contract.json").toString("ascii"));
    const declareResponseTest = await account0.declare({ contract: compiledTest });

    console.log('Test Contract Class Hash =', declareResponseTest.class_hash);
    ```
    What could be the origin of this signature error?

**Comment by tabaktoni** [comment]
This PR should standardize this issue: https://github.com/0xs34n/starknet.js/pull/637

Regarding Cairo 1 procedure to redeclare will be (tested with devnet):
1. get compiled contract (.json)
```const contract = await provider.getClassAt(testAddress);```
2.  Store it to file to get casm
 ```fs.writeFileSync('./testp.json', json.stringify(contract, undefined, 2));```
3. compile stored file to casm usign cairo
```
cargo run --bin starknet-sierra-compile -- /Users/eedev/Work/ShardLabs/starknet.js/__tests__/testp.json /Users/eedev/Work/ShardLabs/starknet.js/__tests__/testp.casm
```
4. load compiled file
 ```
const casm = json.parse(
        fs.readFileSync(path.resolve(__dirname, `../__tests__/testp.casm`)).toString('ascii')
);
```
5. and than declare
```
      const dec = await account.declare({
        contract: contract,
        casm: csm,
      })
```

**Comment by PhilippeR26** [comment]
Nice. I will make some tests Cairo 0 / 1.

**Comment by PhilippeR26** [comment]
Waiting #637 merged ; contract source types are still messy in v5.13.

**Comment by PhilippeR26** [comment]
#637  is now merged.
I made some tests.
Output of `getClassAt` and `getClassByHash` has type : `ContractClassResponse`, whatever the Cairo version : Nice.
`ContractClassResponse.abi` has type : `Abi` whatever the Cairo version : Nice too.
Unfortunately, it generates a new problem :
For `declare` and `declareAndDeploy`, the compiler is waiting `CompiledContract` type, but `ContractClassResponse` is not accepted :
![image](https://github.com/0xs34n/starknet.js/assets/81040730/54b44cf0-bfc8-4310-a752-acc6f585ae2e)
Error message :
`
Impossible to assign type 'ContractClassResponse'  to type 'string | CompiledContract'.ts(2322)
`
How to proceed?

**Comment by PhilippeR26** [comment]
Hello,
I tested the modification made by commit 00b7bdf756c516a1905ce1ecc9036472800989af, and I made some additional investigations :
With the contract of DAI token (Cairo 0 contract on Mainnet) : 0x00da114221cb83fa859dbdb4c44beeaa0bb37c7537ad5ae66fe5e0efd20e6eb3
Let say its classHash is A (to simplify the explanation), read from Starkscan.
Read and decompress contract from : 
- Alchemy : calculated class hash is B.
- My local Pathfinder node :  calculated class hash is B.
- Sequencer Mainnet :  calculated class hash is C.

So, nobody is able to recover the original compiled contract.... 😡

Then I tried to declare in Devnet these compiled contracts.
With just the compiled contract as parameter, it's not working in both DevnetSequencer and DevnetRpc.
If I add in the declare command the class hash A as parameter, it works in all combinations :

- Read from MainnetRpc, Write with DevnetRpc + class Hash A
- Read from MainnetSequencer, Write with DevnetRpc + class Hash A
- Read from MainnetRpc, Write with DevnetSequencer + class Hash A
- Read from MainnetSequencer, Write with DevnetSequencer + class Hash A

Then I tried to read and decompress the contract declared in Devnet.
- With Devnet Rpc, the getClassAt contract do not generate the class Hash A.
- With Devnet Sequencer, the getClassAt contract **generate** the class Hash A.

In summary, the following programs are not able to provide a correct Cairo0 compiled contract :

- Pathfinder
- Starknet Mainnet sequencer.
- Starknet-devnet with Rpc connection.

The only one program to work properly is Starknet-devnet with sequencer access.

**Conclusion :**
- **If you work with a Cairo 1 contract, it works perfectly in all cases.**
- **If you work with a Cairo 0 contract, you need as input the classHash, to decompress the program, and to declare with this class hash as parameter.**

**Comment by PhilippeR26** [comment]
Example for a Cairo 1 contract, from Mainnet to Devnet :
```typescript
const account0 = new Account(providerDevnet, accountAddress0, privateKey0);
const addrErc20Cairo1 = "0x060cf64cf9edfc1b16ec903cee31a2c21680ee02fc778225dacee578c303806a";
const classHashErc20Cairo1 = "0x021b9a01b8f17682f92c18ac4fa08ea90b371f07e27cd08de8451892546d341f";

const compiledSierra1: ContractClassResponse = await providerMainnetRpcLocNetwork.getClassAt(addrErc20Cairo1); // with sequencer or rpc
fs.writeFileSync("./c1_Sierra.json", json.stringify(compiledSierra1, undefined, 2));
const compiledCasm1 = await providerMainnetSequencer.getCompiledClassByClassHash(classHashErc20Cairo1); // only with a sequencer
fs.writeFileSync("./c1_Casm.json", json.stringify(compiledCasm1, undefined, 2));

const compiledC12 = compiledSierra1 as CompiledContract;
const deployResponse1 = await account0.declare({ contract: compiledC12, casm: compiledCasm1 });
```

**Comment by PhilippeR26** [comment]
Example for a Cairo 0 contract, from Mainnet to Devnet :
```typescript
const account0 = new Account(providerDevnet, accountAddress0, privateKey0);
const classHashErc20Cairo0 = "0x01cb96b938da26c060d5fd807eef8b580c49490926393a5eeb408a89f84b9b46";
const addrErc20Cairo0Mainnet = "0x00da114221cb83fa859dbdb4c44beeaa0bb37c7537ad5ae66fe5e0efd20e6eb3";

const compressedContract0: ContractClassResponse = await providerMainnetSequencer.getClassAt(addrErc20Cairo0Mainnet);
const compiledContract0: LegacyCompiledContract = contractClassResponseToLegacyCompiledContract(compressedContract0);

const deployResponse0 = await account0.declareAndDeploy({ contract: compiledContract0,classHash:classHashErc20Cairo0 ,constructorCalldata:[0]});
```
